ee/clickhouse/queries/cohort_query.py

```diff
@@ -0,0 +1,573 @@
+
+        prev_alias: Optional[str] = None
+        fields = ""
+        for idx, (subq_query, subq_alias) in enumerate(filtered_queries):
+            if idx == 0:
+                q += f"({subq_query}) {subq_alias}"
+                fields = f"{subq_alias}.person_id"
+            elif prev_alias:  # can't join without a previous alias
+                q = f"{q} {full_outer_join_query(subq_query, subq_alias, f'{subq_alias}.person_id', f'{prev_alias}.person_id')}"
+                fields = if_condition(
+                    f"{prev_alias}.person_id = '00000000-0000-0000-0000-000000000000'",
+                    f"{subq_alias}.person_id",
+                    f"{fields}",
+                )
+            prev_alias = subq_alias
+
+        return q, fields
+
+    def _get_behavior_subquery(self) -> Tuple[str, Dict[str, Any], str]:
+        #
+        # Get the subquery for the cohort query.
+        #
+
+        query, params = "", {}
+        if self._should_join_behavioral_query:
+
+            _fields = [f"{self.DISTINCT_ID_TABLE_ALIAS}.person_id AS person_id"]
+            _fields.extend(self._fields)
+
+            date_condition, date_params = self._get_date_condition()
+
+            query = f"""
+            SELECT {", ".join(_fields)} FROM events {self.EVENT_TABLE_ALIAS}
+            {self._get_distinct_id_query()}
+            WHERE team_id = %(team_id)s
+            AND event IN %(events)s
+            {date_condition}
+            GROUP BY person_id
+            """
+
+            query, params = (query, {"team_id": self._team_id, "events": self._events, **date_params,})
+
+        return query, params, self.BEHAVIOR_QUERY_ALIAS
+
+    def _get_persons_query(self) -> Tuple[str, Dict[str, Any], str]:
+        query, params = "", {}
+        if self._should_join_persons:
+            person_query, person_params = self._person_query.get_query()
+            person_query = f"SELECT *, id AS person_id FROM ({person_query})"
+
+            query, params = person_query, person_params
+
+        return query, params, self.PERSON_TABLE_ALIAS
+
+    def _get_date_condition(self) -> Tuple[str, Dict[str, Any]]:
+        # TODO: handle as params
+        date_query = ""
+        date_params: Dict[str, Any] = {}
+
+        if self._earliest_time:
+            date_query = f"AND timestamp <= now() AND timestamp >= now()"
+            for i, date in enumerate(self._earliest_time):
+                param = f"earliest_time_{i}"
+                date_query += f" - INTERVAL %({param})s {validate_interval(date[1])}"
+                date_params = {**date_params, param: date[0]}
+
+        return date_query, date_params
+
+    def _check_earliest_date(self, relative_dates: List[Relative_Date]) -> None:
+        if not self._earliest_time:
+            self._earliest_time = relative_dates
+        elif relative_date_is_greater(relative_dates, self._earliest_time):
+            self._earliest_time = relative_dates
```

Comment:
For `restarted_performing_event` and `performed_event_first_time`, the earliest date is "all time", so we should support that concept here. Makes me wonder if the earliest date should be event specific. Might be over optimizing, so just a thought.